charmcraft/parts.py

```diff
@@ -0,0 +1,243 @@
+# Copyright 2021 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# For further info, check https://github.com/canonical/charmcraft
+
+"""Craft-parts setup, lifecycle and plugins."""
+
+import logging
+import os
+import pathlib
+import shlex
+import sys
+from typing import Any, Dict, List, Set
+
+from craft_parts import LifecycleManager, Step, plugins
+from craft_parts.parts import PartSpec
+from craft_parts.errors import PartsError
+from xdg import BaseDirectory  # type: ignore
+
+from charmcraft import charm_builder
+from charmcraft.cmdbase import CommandError
+
+logger = logging.getLogger(__name__)
+
+
+class CharmPluginProperties(plugins.PluginProperties, plugins.PluginModel):
+    """Properties used in charm building."""
+
+    source: str = ""
+    # TODO: add charm plugin properties
+
+    @classmethod
+    def unmarshal(cls, data: Dict[str, Any]):
+        """Populate charm properties from the part specification.
+
+        :param data: A dictionary containing part properties.
+
+        :return: The populated plugin properties data object.
+
+        :raise pydantic.ValidationError: If validation fails.
+        """
+        plugin_data = plugins.extract_plugin_properties(
+            data, plugin_name="charm", required=["source"]
+        )
+        return cls(**plugin_data)
+
+
+class CharmPlugin(plugins.Plugin):
+    """Build the charm and prepare for packing.
+
+    The craft-parts charm plugin prepares the charm payload for packing. Common
+    plugin and source keywords can be used.
+
+    Extra files to be included in the charm payload must be listed under
+    the ``prime`` file filter.
+    """
+
+    properties_class = CharmPluginProperties
+
+    @classmethod
+    def get_build_snaps(cls) -> Set[str]:
+        """Return a set of required snaps to install in the build environment."""
+        return set()
+
+    def get_build_packages(self) -> Set[str]:
+        """Return a set of required packages to install in the build environment."""
+        return {
+            "python3-pip",
```

Comment:
you can now remove the manual install of these done for the craft-providers integration

Reply:
my comment was the other way around :-) Keep them here where they belong, and remove them from https://github.com/canonical/charmcraft/blob/7cdbeeeb569470b919e68b6c4212206b1e016649/charmcraft/providers.py#L324

Reply:
Ah, I misread "craft-providers" there and parsed it as "craft-parts" for some reason. I guess I'm tired, fixing it.